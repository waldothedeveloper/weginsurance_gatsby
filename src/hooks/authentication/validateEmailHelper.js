/* eslint-disable no-useless-escape */
export const validEmailRegex = RegExp(
  /^(([^<>()\[\]\.,;:\s@\"]+(\.[^<>()\[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i
);
