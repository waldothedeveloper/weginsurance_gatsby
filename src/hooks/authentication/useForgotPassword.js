import { getAuth, sendPasswordResetEmail } from "firebase/auth";
import { useEffect, useState } from "react";

import { app } from "../../utils/firebaseConfig";
import { navigate } from "gatsby";
import { validEmailRegex } from "./validateEmailHelper";

const auth = getAuth(app);
auth.languageCode = `es`;

export const useForgotPassword = () => {
  const [email, setEmail] = useState(``);
  const [validationErrors, setValidationErrors] = useState(null);
  const [authErrors, setAuthErrors] = useState(``);

  useEffect(() => {
    if (!validEmailRegex.test(email)) {
      setValidationErrors(
        `El correo electronico debe contener un formato valido`
      );
    } else {
      setValidationErrors(``);
    }
  }, [email]);

  const handleChange = (event) => {
    const { value, name } = event.target;

    if (name === `email`) {
      setEmail(value);
    }
  };

  const handleSubmit = (event) => {
    if (event) event.preventDefault();

    //
    sendPasswordResetEmail(auth, email)
      .then(() => {
        // Password reset email sent!
        navigate(`/auth/email_sent_successfully`, { state: { email: email } });
      })
      .catch((error) => {
        const { code } = error;

        switch (code) {
          case `auth/user-not-found`:
            setAuthErrors(() => `El usuario o correo electronico no existe.`);
            break;
          default:
            setAuthErrors(
              () =>
                `Ha ocurrido un error inesperado. Intentelo de nuevo mas tarde.`
            );
            break;
        }
      });
  };

  return {
    handleSubmit,
    validationErrors,
    email,
    handleChange,
    authErrors,
    setAuthErrors,
  };
};
