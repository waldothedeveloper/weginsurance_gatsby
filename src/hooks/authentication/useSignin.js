import { useEffect, useState } from "react";

import { navigate } from "gatsby";
import { signInWithEmailAndPassword } from "firebase/auth";
import { useAuthState } from "../../components/shared/auth";
import { validEmailRegex } from "./validateEmailHelper";

export const useSignIn = () => {
  const { firebaseAuth } = useAuthState();
  const [email, setEmail] = useState(``);
  const [password, setPassword] = useState(``);
  const [validationErrors, setValidationErrors] = useState(null);
  const [signInErrors, setSignInErrors] = useState(``);

  useEffect(() => {
    if (email.length > 0 && !validEmailRegex.test(email)) {
      setValidationErrors(
        `El correo electronico debe contener un formato valido`
      );
    } else {
      setValidationErrors(``);
    }
  }, [email]);

  const handleChange = (event) => {
    const { value, name } = event.target;

    if (name === `email`) {
      setEmail(value);
    }
    if (name === `password`) setPassword(value);
  };

  const handleSubmit = (event) => {
    if (event) event.preventDefault();

    signInWithEmailAndPassword(firebaseAuth, email, password)
      .then(() => {
        // ! redirect to admin page
        navigate(`/admin/dashboard`);
      })
      .catch((error) => {
        const { code } = error;
        // console.log(`code: `, code);

        switch (code) {
          case `auth/wrong-password`:
            setSignInErrors(`La contraseña es incorrecta.`);

            break;
          case `auth/user-not-found`:
            setSignInErrors(`El usuario es incorrecto o no existe.`);

            break;
          case `auth/too-many-requests`:
            setSignInErrors(
              `Demasiados intentos fallidos, intente de nuevo mas tarde.`
            );

            break;
          case `auth/invalid-password`:
            setSignInErrors(
              `La contraseña es muy corta. Debe contener al menos 6 caracteres.`
            );

            break;
          default:
            setSignInErrors(
              `Ha ocurrido un error inesperado. Intentelo de nuevo mas tarde.`
            );

            break;
        }
      });
  };

  return {
    handleSubmit,
    validationErrors,
    email,
    password,
    handleChange,
    signInErrors,
    setSignInErrors,
  };
};
