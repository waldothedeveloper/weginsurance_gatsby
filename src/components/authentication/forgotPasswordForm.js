import { LockClosedIcon, LockOpenIcon } from "@heroicons/react/solid";

import { Link } from "gatsby";
import Notifications from "../notifications";
import React from "react";
import { StaticImage } from "gatsby-plugin-image";
import { useForgotPassword } from "../../hooks/authentication/useForgotPassword";

export const ForgotPasswordForm = () => {
  const {
    handleSubmit,
    error,
    email,
    handleChange,
    authErrors,
    openCloseNotification,
    setOpenCloseNotification,
  } = useForgotPassword();
  return (
    <>
      <div className="flex min-h-full items-center justify-center py-12 px-4 sm:px-6 lg:px-8">
        <div className="w-full max-w-md space-y-8">
          <div className="flex flex-col items-center justify-center">
            <StaticImage
              className="mx-auto h-24 w-24 rounded-full"
              placeholder="blurred"
              layout="constrained"
              src="../../images/weginsurance_logo.jpeg"
              alt="weginsurance logo"
            />

            <h2 className="mt-6 text-center text-3xl font-extrabold text-gray-900">
              Recuperar la cuenta
            </h2>
            <p className="mt-2 text-center text-sm text-gray-600">
              Escriba el correo para recibir un link para recuperar la cuenta
            </p>
          </div>

          <form className="space-y-6">
            <div>
              <label
                htmlFor="email"
                className="block text-sm font-medium text-gray-700"
              >
                Correo Electronico
              </label>
              <div className="mt-1">
                <input
                  id="email-address"
                  data-testid="email-address"
                  value={email || ``}
                  onChange={handleChange}
                  name="email"
                  type="email"
                  autoComplete="email"
                  className={
                    error && email.length > 0
                      ? `block w-full appearance-none rounded-md border border-red-500 px-3 py-2 text-red-500 placeholder-red-500 shadow-sm focus:border-red-500 focus:outline-none focus:ring-red-500 sm:text-sm`
                      : `block w-full appearance-none rounded-md border border-gray-300 px-3 py-2 placeholder-gray-400 shadow-sm focus:border-sky-500 focus:outline-none focus:ring-sky-500 sm:text-sm`
                  }
                />
                <p className="mt-2 text-sm text-red-600" id="email-error">
                  {error && email.length > 0 && error}
                </p>
              </div>
            </div>

            <div>
              <button
                data-testid="submit-button"
                disabled={error ?? false}
                onClick={handleSubmit}
                className={
                  error
                    ? `group relative flex w-full justify-center rounded-md border border-transparent bg-gray-400 px-4 py-2 text-sm font-medium text-white  focus:outline-none focus:ring-2 focus:ring-gray-500 focus:ring-offset-2`
                    : `group relative flex w-full justify-center rounded-md border border-transparent bg-sky-600 py-2 px-4 text-sm font-medium text-white hover:bg-sky-700 focus:outline-none focus:ring-2 focus:ring-sky-500 focus:ring-offset-2`
                }
              >
                <span className="absolute inset-y-0 left-0 flex items-center pl-3">
                  {error ? (
                    <LockClosedIcon
                      className="h-5 w-5 text-gray-500"
                      aria-hidden="true"
                    />
                  ) : (
                    <LockOpenIcon
                      className="h-5 w-5 text-sky-500 group-hover:text-sky-400"
                      aria-hidden="true"
                    />
                  )}
                </span>
                Enviar correo
              </button>
            </div>
          </form>
          <div className="flex items-center justify-end">
            <div className="text-sm">
              <Link
                data-testid="sign-in-link"
                to="/auth/login"
                className="font-medium text-sky-600 hover:text-sky-500"
              >
                Regresar a la pagina de inicio de sesi√≥n
              </Link>
            </div>
          </div>
        </div>
      </div>
      {authErrors && (
        <Notifications
          authErrors={authErrors}
          openCloseNotification={openCloseNotification}
          setOpenCloseNotification={setOpenCloseNotification}
        />
      )}
    </>
  );
};
