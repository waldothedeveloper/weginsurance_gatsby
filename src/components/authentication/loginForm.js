import { LockClosedIcon, LockOpenIcon } from "@heroicons/react/solid";

import { Link } from "gatsby";
import Notifications from "../notifications";
import React from "react";
import { StaticImage } from "gatsby-plugin-image";
import { useSignIn } from "../../hooks/authentication/useSignin";

export const LoginForm = () => {
  const {
    handleSubmit,
    error,
    email,
    password,
    handleChange,
    authErrors,
    openCloseNotification,
    setOpenCloseNotification,
  } = useSignIn();
  return (
    <>
      <div className="flex min-h-full items-center justify-center py-12 px-4 sm:px-6 lg:px-8">
        <div className="w-full max-w-md space-y-8">
          <div className="flex flex-col items-center justify-center">
            <StaticImage
              className="mx-auto h-24 w-24 rounded-full"
              placeholder="blurred"
              layout="constrained"
              src="../../images/weginsurance_logo.jpeg"
              alt="weginsurance logo"
            />

            <h2 className="mt-6 text-center text-3xl font-extrabold text-gray-900">
              Entrar a la cuenta
            </h2>
            <p className="mt-2 text-center text-sm text-gray-600">
              Si tiene problemas para entrar contacte al soporte tecnico
            </p>
          </div>

          <form className="space-y-6">
            <div>
              <label
                htmlFor="email"
                className="block text-sm font-medium text-gray-700"
              >
                Correo Electronico
              </label>
              <div className="mt-1">
                <input
                  id="email-address"
                  value={email || ``}
                  onChange={handleChange}
                  name="email"
                  type="email"
                  autoComplete="email"
                  className={
                    error && email.length > 0
                      ? `block w-full appearance-none rounded-md border border-red-500 px-3 py-2 text-red-500 placeholder-red-500 shadow-sm focus:border-red-500 focus:outline-none focus:ring-red-500 sm:text-sm`
                      : `block w-full appearance-none rounded-md border border-gray-300 px-3 py-2 placeholder-gray-400 shadow-sm focus:border-sky-500 focus:outline-none focus:ring-sky-500 sm:text-sm`
                  }
                />
                <p className="mt-2 text-sm text-red-600" id="email-error">
                  {error && email.length > 0 && error}
                </p>
              </div>
            </div>

            <div>
              <label
                htmlFor="password"
                className="block text-sm font-medium text-gray-700"
              >
                Contraseña
              </label>
              <div className="mt-1">
                <input
                  id="password"
                  value={password || ``}
                  onChange={handleChange}
                  name="password"
                  type="password"
                  autoComplete="current-password"
                  className="block w-full appearance-none rounded-md border border-gray-300 px-3 py-2 placeholder-gray-400 shadow-sm focus:border-sky-500 focus:outline-none focus:ring-sky-500 sm:text-sm"
                />
              </div>
            </div>

            <div className="flex items-center justify-end">
              <div className="text-sm">
                <Link
                  to="/auth/forgot_password"
                  className="font-medium text-sky-600 hover:text-sky-500"
                >
                  Olvido su Contraseña?
                </Link>
              </div>
            </div>
            <div>
              <button
                disabled={(error || password.length === 0) ?? false}
                onClick={handleSubmit}
                className={
                  error || password.length === 0
                    ? `group relative flex w-full justify-center rounded-md border border-transparent bg-gray-400 px-4 py-2 text-sm font-medium text-white  focus:outline-none focus:ring-2 focus:ring-gray-500 focus:ring-offset-2`
                    : `group relative flex w-full justify-center rounded-md border border-transparent bg-sky-600 py-2 px-4 text-sm font-medium text-white hover:bg-sky-700 focus:outline-none focus:ring-2 focus:ring-sky-500 focus:ring-offset-2`
                }
              >
                <span className="absolute inset-y-0 left-0 flex items-center pl-3">
                  {error || password.length === 0 ? (
                    <LockClosedIcon
                      className="h-5 w-5 text-gray-500"
                      aria-hidden="true"
                    />
                  ) : (
                    <LockOpenIcon
                      className="h-5 w-5 text-sky-500 group-hover:text-sky-400"
                      aria-hidden="true"
                    />
                  )}
                </span>
                Entrar
              </button>
            </div>
          </form>
        </div>
      </div>
      {authErrors && (
        <Notifications
          authErrors={authErrors}
          openCloseNotification={openCloseNotification}
          setOpenCloseNotification={setOpenCloseNotification}
        />
      )}
    </>
  );
};
