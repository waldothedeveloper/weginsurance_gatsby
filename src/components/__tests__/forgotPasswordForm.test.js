import { render, screen } from "@testing-library/react";

// import EmailSentSuccessfully from "../../pages/auth/emailSentSuccessfully";
import { ForgotPasswordForm } from "../authentication/forgotPasswordForm";
import React from "react";
import { navigate } from "gatsby";
import userEvent from "@testing-library/user-event";

jest.mock(`gatsby-plugin-image`, () => {
  const React = require(`react`);
  const plugin = jest.requireActual(`gatsby-plugin-image`);

  const mockImage = ({ imgClassName, ...props }) =>
    React.createElement(`img`, { ...props, className: imgClassName });

  const mockPlugin = {
    ...plugin,
    GatsbyImage: jest.fn().mockImplementation(mockImage),
    StaticImage: jest.fn().mockImplementation(mockImage),
  };

  return mockPlugin;
});

const user = userEvent.setup();

describe(`Forgot Password Form`, () => {
  it(`renders correctly`, () => {
    const container = render(<ForgotPasswordForm />);

    expect(container).toMatchSnapshot();
  });

  it(`submit button should be disabled if there's the input field is blank`, () => {
    render(<ForgotPasswordForm />);
    const submitButton = screen.getByTestId(`submit-button`);

    expect(submitButton).toBeDisabled();
  });

  it(`submit button should be disabled if an invalid email is typed`, async () => {
    render(<ForgotPasswordForm />);
    const inputField = screen.getByTestId(`email-address`);
    await user.type(inputField, `some invalid email`);
    const submitButton = screen.getByTestId(`submit-button`);

    expect(submitButton).toBeDisabled();
  });

  it(`Submit button should NOT be disabled if a valid email address is used`, async () => {
    render(<ForgotPasswordForm />);
    const button = screen.getByTestId(`submit-button`);
    const inputEl = screen.getByTestId(`email-address`);
    await user.type(inputEl, `test2@gmail.com`);

    expect(inputEl.value).toBe(`test2@gmail.com`);
    expect(button).toBeEnabled();
  });

  it(`navigates back to /auth/login when Gatsby Link (anchor) is clicked`, async () => {
    render(<ForgotPasswordForm />);
    const link = screen.getByTestId(`sign-in-link`);

    await user.click(link);
    await navigate(`/auth/login`);

    expect(navigate).toHaveBeenCalledTimes(1);
    expect(navigate).toHaveBeenCalledWith(`/auth/login`);
  });
});
