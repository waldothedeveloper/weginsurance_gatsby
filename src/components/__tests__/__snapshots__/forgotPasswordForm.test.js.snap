// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forgot Password Form renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      
      <div
        class="flex min-h-full items-center justify-center py-12 px-4 sm:px-6 lg:px-8"
      >
        <div
          class="w-full max-w-md space-y-8"
        >
          <div
            class="flex flex-col items-center justify-center"
          >
            <img
              alt="weginsurance logo"
              layout="constrained"
              placeholder="blurred"
              src="../../images/weginsurance_logo.jpeg"
            />
            <h2
              class="mt-6 text-center text-3xl font-extrabold text-gray-900"
            >
              Recuperar la cuenta
            </h2>
            <p
              class="mt-2 text-center text-sm text-gray-600"
            >
              Escriba el correo para recibir un link para recuperar la cuenta
            </p>
          </div>
          <form
            class="space-y-6"
          >
            <div>
              <label
                class="block text-sm font-medium text-gray-700"
                for="email"
              >
                Correo Electronico
              </label>
              <div
                class="mt-1"
              >
                <input
                  autocomplete="email"
                  class="block w-full appearance-none rounded-md border border-gray-300 px-3 py-2 placeholder-gray-400 shadow-sm focus:border-sky-500 focus:outline-none focus:ring-sky-500 sm:text-sm"
                  data-testid="email-address"
                  id="email-address"
                  name="email"
                  type="email"
                  value=""
                />
                <p
                  class="mt-2 text-sm text-red-600"
                  id="email-error"
                />
              </div>
            </div>
            <div>
              <button
                class="group relative flex w-full justify-center rounded-md border border-transparent bg-gray-400 px-4 py-2 text-sm font-medium text-white  focus:outline-none focus:ring-2 focus:ring-gray-500 focus:ring-offset-2"
                data-testid="submit-button"
                disabled=""
              >
                <span
                  class="absolute inset-y-0 left-0 flex items-center pl-3"
                >
                  <svg
                    aria-hidden="true"
                    class="h-5 w-5 text-gray-500"
                    fill="currentColor"
                    viewBox="0 0 20 20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M5 9V7a5 5 0 0110 0v2a2 2 0 012 2v5a2 2 0 01-2 2H5a2 2 0 01-2-2v-5a2 2 0 012-2zm8-2v2H7V7a3 3 0 016 0z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
                Enviar correo
              </button>
            </div>
          </form>
          <div
            class="flex items-center justify-end"
          >
            <div
              class="text-sm"
            >
              <a
                class="font-medium text-sky-600 hover:text-sky-500"
                data-testid="sign-in-link"
                href="/auth/login"
              >
                Regresar a la pagina de inicio de sesión
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    
    <div
      class="flex min-h-full items-center justify-center py-12 px-4 sm:px-6 lg:px-8"
    >
      <div
        class="w-full max-w-md space-y-8"
      >
        <div
          class="flex flex-col items-center justify-center"
        >
          <img
            alt="weginsurance logo"
            layout="constrained"
            placeholder="blurred"
            src="../../images/weginsurance_logo.jpeg"
          />
          <h2
            class="mt-6 text-center text-3xl font-extrabold text-gray-900"
          >
            Recuperar la cuenta
          </h2>
          <p
            class="mt-2 text-center text-sm text-gray-600"
          >
            Escriba el correo para recibir un link para recuperar la cuenta
          </p>
        </div>
        <form
          class="space-y-6"
        >
          <div>
            <label
              class="block text-sm font-medium text-gray-700"
              for="email"
            >
              Correo Electronico
            </label>
            <div
              class="mt-1"
            >
              <input
                autocomplete="email"
                class="block w-full appearance-none rounded-md border border-gray-300 px-3 py-2 placeholder-gray-400 shadow-sm focus:border-sky-500 focus:outline-none focus:ring-sky-500 sm:text-sm"
                data-testid="email-address"
                id="email-address"
                name="email"
                type="email"
                value=""
              />
              <p
                class="mt-2 text-sm text-red-600"
                id="email-error"
              />
            </div>
          </div>
          <div>
            <button
              class="group relative flex w-full justify-center rounded-md border border-transparent bg-gray-400 px-4 py-2 text-sm font-medium text-white  focus:outline-none focus:ring-2 focus:ring-gray-500 focus:ring-offset-2"
              data-testid="submit-button"
              disabled=""
            >
              <span
                class="absolute inset-y-0 left-0 flex items-center pl-3"
              >
                <svg
                  aria-hidden="true"
                  class="h-5 w-5 text-gray-500"
                  fill="currentColor"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M5 9V7a5 5 0 0110 0v2a2 2 0 012 2v5a2 2 0 01-2 2H5a2 2 0 01-2-2v-5a2 2 0 012-2zm8-2v2H7V7a3 3 0 016 0z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              Enviar correo
            </button>
          </div>
        </form>
        <div
          class="flex items-center justify-end"
        >
          <div
            class="text-sm"
          >
            <a
              class="font-medium text-sky-600 hover:text-sky-500"
              data-testid="sign-in-link"
              href="/auth/login"
            >
              Regresar a la pagina de inicio de sesión
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
